ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -96,18 +103,36 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
    * @param size the number of bytes to upload.
    * @throws CloudStorageException if the upload failed.
    */
-  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size) throws CloudStorageException {
-    boolean performUpload = messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !messageInfo.isDeleted();
-    // May need to encrypt buffer before upload
-    if (performUpload) {
+  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
+      throws CloudStorageException, IOException {
+    if (shouldUpload(messageInfo)) {
       BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      if (cloudConfig.vcrRequireEncryption && !BlobId.isEncrypted(blobId.getID())) {
+        // TODO: encrypt on demand before uploading, skip for now
+        return;
+      }
       CloudBlobMetadata blobMetadata =
           new CloudBlobMetadata(blobId, messageInfo.getOperationTimeMs(), messageInfo.getExpirationTimeInMs(),
               messageInfo.getSize());
       cloudDestination.uploadBlob(blobId, size, blobMetadata, new ByteBufferInputStream(messageBuf));
     }
   }
 
+  /**
+   * Utility to decide whether a blob should be uploaded.
+   * @param messageInfo The {@link MessageInfo} containing the blob metadata.
+   * @return {@code true} is the blob should be upload, {@code false} otherwise.
+   */
+  private boolean shouldUpload(MessageInfo messageInfo) {
+    if (messageInfo.isDeleted()) {
+      return false;
+    }
+
+    // expiration time above threshold
+    return (messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time
```

Comment:
How does this behave with TTLUpdated blobs? Double check that expirationtimeinms is updated when a TTLUpdate happens

Reply:
We'll test it.